### DISCUSSION TITLE
Fine tuning training errors.

### DISCUSSION THREAD

**Discussion by rmcpantoja** [coqui-ai/TTS#2324] [General Q&A]
Hello,
I'm trying to fine tune a model using a dataset of an hour and a half, but during the final part of the training I have problems, but first I will explain:
The model I'm trying to work with is: tts_models/es/css10/vits, which was trained with 🐸TTS version 0.8 and therefore:

1. I've reverted the TTS repository back to the commit that corresponds to the version 0.8.
2. I've taken the file [config.json](https://cdn.discordapp.com/attachments/1037421693569871942/1071527297489846396/config.json) of this pre-trained model already downloaded, adjusting it to the paths of my dataset and some training settings.
3. When training with python train TTS, I get the following error: 'VITS' has no attribute 'disc'. All of these steps can be reproduced in a [colab notebook](https://colab.research.google.com/drive/18lsSAHAttlzyuztRI_5UVTntzfAOj9F8) (with the machines offered by Google).

Is there something wrong in the configuration? since I have searched and I can't find anything related to "disc", not even the pretrained model had such a configuration. The only thing I found is rl_disc.
Is there any solution for this? Thank you very much in advance.

**Comment by erogol** [comment]
The released model has no discriminator to keep its memory footprint smaller. You need to set `init_discriminator` false and try loading the model.

**Reply by rmcpantoja** [reply]
> The released model has no discriminator to keep its memory footprint smaller. You need to set `init_discriminator` false and try loading the model.

Hi @erogol init_discriminator was always set to False in my config.json, since I took the configs from the pretrained model just addjusting it to my dataset. Here is the config, taking into account that I moved the repository back to version 0.8 also due to changes in the values for the configs:
# config:
{
    "output_path": "/content/drive/MyDrive/TTSWork",
    "logger_uri": null,
    "run_name": "FT",
    "project_name": null,
    "run_description": "",
    "print_step": 25,
    "plot_step": 100,
    "model_param_stats": false,
    "wandb_entity": null,
    "dashboard_logger": "tensorboard",
    "log_model_step": 1000,
    "save_step": 10000,
    "save_n_checkpoints": 5,
    "save_checkpoints": true,
    "save_all_best": false,
    "save_best_after": 10000,
    "target_loss": null,
    "print_eval": false,
    "test_delay_epochs": -1,
    "run_eval": true,
    "distributed_backend": "nccl",
    "distributed_url": "tcp://localhost:54321",
    "mixed_precision": true,
    "epochs": 1000,
    "batch_size": 32,
    "eval_batch_size": 4,
    "grad_clip": [
        1000,
        1000
    ],
    "scheduler_after_epoch": true,
    "lr": 0.0002,
    "optimizer": "AdamW",
    "optimizer_params": {
        "betas": [
            0.8,
            0.99
        ],
        "eps": 1e-09,
        "weight_decay": 0.01
    },
    "lr_scheduler": "",
    "lr_scheduler_params": {},
    "use_grad_scaler": false,
    "cudnn_enable": true,
    "cudnn_deterministic": false,
    "cudnn_benchmark": false,
    "training_seed": 54321,
    "model": "vits",
    "num_loader_workers": 8,
    "num_eval_loader_workers": 8,
    "use_noise_augment": false,
    "audio": {
        "fft_size": 1024,
        "win_length": 1024,
        "hop_length": 256,
        "frame_shift_ms": null,
        "frame_length_ms": null,
        "stft_pad_mode": "reflect",
        "sample_rate": 22050,
        "resample": false,
        "preemphasis": 0.0,
        "ref_level_db": 20,
        "do_sound_norm": false,
        "log_func": "np.log",
        "do_trim_silence": true,
        "trim_db": 45,
        "do_rms_norm": false,
        "db_level": null,
        "power": 1.5,
        "griffin_lim_iters": 60,
        "num_mels": 80,
        "mel_fmin": 0.0,
        "mel_fmax": null,
        "spec_gain": 1.0,
        "do_amp_to_db_linear": false,
        "do_amp_to_db_mel": true,
        "pitch_fmax": 640.0,
        "pitch_fmin": 0.0,
        "signal_norm": true,
        "min_level_db": -100,
        "symmetric_norm": true,
        "max_norm": 4.0,
        "clip_norm": true,
        "stats_path": null
    },
    "use_phonemes": false,
    "phonemizer": null,
    "phoneme_language": null,
    "compute_input_seq_cache": true,
    "text_cleaner": "multilingual_cleaners",
    "enable_eos_bos_chars": false,
    "test_sentences_file": "",
    "phoneme_cache_path": null,
    "characters": {
        "characters_class": "TTS.tts.utils.text.characters.Graphemes",
        "vocab_dict": null,
        "pad": "<PAD>",
        "eos": "<EOS>",
        "bos": "<BOS>",
        "blank": "<BLNK>",
        "characters": "abcdefghijklmnopqrstuvwxyz\u00af\u00b7\u00df\u00e0\u00e1\u00e2\u00e3\u00e4\u00e6\u00e7\u00e8\u00e9\u00ea\u00eb\u00ec\u00ed\u00ee\u00ef\u00f1\u00f2\u00f3\u00f4\u00f5\u00f6\u00f9\u00fa\u00fb\u00fc\u00ff\u0101\u0105\u0107\u0113\u0119\u011b\u012b\u0131\u0142\u0144\u014d\u0151\u0153\u015b\u016b\u0171\u017a\u017c\u01ce\u01d0\u01d2\u01d4\u0430\u0431\u0432\u0433\u0434\u0435\u0436\u0437\u0438\u0439\u043a\u043b\u043c\u043d\u043e\u043f\u0440\u0441\u0442\u0443\u0444\u0445\u0446\u0447\u0448\u0449\u044a\u044b\u044c\u044d\u044e\u044f\u0451\u0454\u0456\u0457\u0491",
        "punctuations": "¡!'(),-.:;¿? ",
        "phonemes": null,
        "is_unique": true,
        "is_sorted": true
    },
    "add_blank": true,
    "batch_group_size": 0,
    "loss_masking": null,
    "sort_by_audio_len": false,
    "min_audio_len": 32768,
    "max_audio_len": 132300,
    "min_text_len": 1,
    "max_text_len": Infinity,
    "compute_f0": false,
    "compute_linear_spec": true,
    "precompute_num_workers": 12,
    "start_by_longest": false,
    "datasets": [
        {
            "name": "css10",
            "path": "/content/TTS/data",
            "meta_file_train": "metadata.csv",
            "ignored_speakers": null,
            "language": "es",
            "meta_file_val": "",
            "meta_file_attn_mask": ""
        }
    ],
    "test_sentences": [
        [
            "Un arco\u00edris\u200b o arco iris es un fen\u00f3meno \u00f3ptico y meteorol\u00f3gico que causa la aparici\u00f3n en la atm\u00f3sfera terrestre de un arco multicolor.",
            "css10",
            null,
            "es"
        ]
    ],
    "eval_split_max_size": null,
    "eval_split_size": 0.01,
    "use_speaker_weighted_sampler": false,
    "speaker_weighted_sampler_alpha": 1.0,
    "use_language_weighted_sampler": true,
    "language_weighted_sampler_alpha": 1.0,
    "model_args": {
        "num_chars": 129,
        "out_channels": 513,
        "spec_segment_size": 32,
        "hidden_channels": 192,
        "hidden_channels_ffn_text_encoder": 768,
        "num_heads_text_encoder": 2,
        "num_layers_text_encoder": 6,
        "kernel_size_text_encoder": 3,
        "dropout_p_text_encoder": 0.1,
        "dropout_p_duration_predictor": 0.5,
        "kernel_size_posterior_encoder": 5,
        "dilation_rate_posterior_encoder": 1,
        "num_layers_posterior_encoder": 16,
        "kernel_size_flow": 5,
        "dilation_rate_flow": 1,
        "num_layers_flow": 4,
        "resblock_type_decoder": "2",
        "resblock_kernel_sizes_decoder": [
            3,
            5,
            7
        ],
        "resblock_dilation_sizes_decoder": [
            [
                1,
                2
            ],
            [
                2,
                6
            ],
            [
                3,
                12
            ]
        ],
        "upsample_rates_decoder": [
            8,
            8,
            4
        ],
        "upsample_initial_channel_decoder": 256,
        "upsample_kernel_sizes_decoder": [
            16,
            16,
            8
        ],
        "use_sdp": true,
        "noise_scale": 1.0,
        "inference_noise_scale": 0.3,
        "length_scale": 1,
        "noise_scale_dp": 1.0,
        "inference_noise_scale_dp": 0.3,
        "max_inference_len": null,
        "init_discriminator": false,
        "use_spectral_norm_disriminator": false,
        "use_speaker_embedding": true,
        "num_speakers": 0,
        "speakers_file": "/content/TTS/tts_models--es--css10--vits/speaker_ids.json",
        "d_vector_file": null,
        "speaker_embedding_channels": 256,
        "use_d_vector_file": false,
        "d_vector_dim": 0,
        "detach_dp_input": true,
        "use_language_embedding": true,
        "embedded_language_dim": 4,
        "num_languages": 1,
        "language_ids_file": "/content/TTS/tts_models--es--css10--vits/language_ids.json",
        "use_speaker_encoder_as_loss": false,
        "speaker_encoder_config_path": "",
        "speaker_encoder_model_path": "",
        "condition_dp_on_speaker": true,
        "freeze_encoder": false,
        "freeze_DP": false,
        "freeze_PE": false,
        "freeze_flow_decoder": false,
        "freeze_waveform_decoder": false
    },
    "lr_gen": 0.0002,
    "lr_disc": 0.0002,
    "lr_scheduler_gen": "ExponentialLR",
    "lr_scheduler_gen_params": {
        "gamma": 0.999875,
        "last_epoch": -1
    },
    "lr_scheduler_disc": "ExponentialLR",
    "lr_scheduler_disc_params": {
        "gamma": 0.999875,
        "last_epoch": -1
    },
    "kl_loss_alpha": 1.0,
    "disc_loss_alpha": 1.0,
    "gen_loss_alpha": 1.0,
    "feat_loss_alpha": 1.0,
    "mel_loss_alpha": 45.0,
    "dur_loss_alpha": 1.0,
    "speaker_encoder_loss_alpha": 1.0,
    "return_wav": true,
    "r": 1,
    "num_speakers": 1,
    "use_speaker_embedding": true,
    "speakers_file": "/content/TTS/tts_models--es--css10--vits/speaker_ids.json",
    "speaker_embedding_channels": 256,
    "language_ids_file": "/content/TTS/tts_models--es--css10--vits/language_ids.json",
    "use_language_embedding": true,
    "use_d_vector_file": false,
    "d_vector_file": null,
    "d_vector_dim": 0
}

# training log:
Command: !CUDA_VISIBLE_DEVICES="0" python TTS/bin/train_tts.py --config_path "/content/config.json" --restore_path "/content/TTS/tts_models--es--css10--vits/model_file.pth.tar"
 | > Found 535 files in /content/TTS/data
 > Using model: vits
 > Setting up Audio Processor...
 | > sample_rate:22050
 | > resample:False
 | > num_mels:80
 | > log_func:np.log10
 | > min_level_db:0
 | > frame_shift_ms:None
 | > frame_length_ms:None
 | > ref_level_db:None
 | > fft_size:1024
 | > power:None
 | > preemphasis:0.0
 | > griffin_lim_iters:None
 | > signal_norm:None
 | > symmetric_norm:None
 | > mel_fmin:0
 | > mel_fmax:None
 | > pitch_fmin:None
 | > pitch_fmax:None
 | > spec_gain:20.0
 | > stft_pad_mode:reflect
 | > max_norm:1.0
 | > clip_norm:True
 | > do_trim_silence:False
 | > trim_db:60
 | > do_sound_norm:False
 | > do_amp_to_db_linear:True
 | > do_amp_to_db_mel:True
 | > do_rms_norm:False
 | > db_level:None
 | > stats_path:None
 | > base:10
 | > hop_length:256
 | > win_length:1024
 > initialization of speaker-embedding layers.
 > initialization of language-embedding layers.
 > Training Environment:
 | > Current device: 0
 | > Num. of GPUs: 1
 | > Num. of CPUs: 2
 | > Num. of Torch Threads: 1
 | > Torch seed: 54321
 | > Torch CUDNN: True
 | > Torch CUDNN deterministic: False
 | > Torch CUDNN benchmark: False
 > Start Tensorboard: tensorboard --logdir=/content/drive/MyDrive/TTSWork/FT-February-06-2023_03+17PM-2c9f00a8
Traceback (most recent call last):
  File "TTS/bin/train_tts.py", line 71, in <module>
    main()
  File "TTS/bin/train_tts.py", line 58, in main
    trainer = Trainer(
  File "/usr/local/lib/python3.8/dist-packages/trainer/trainer.py", line 489, in __init__
    self.optimizer = self.get_optimizer(self.model, self.config)
  File "/usr/local/lib/python3.8/dist-packages/trainer/trainer.py", line 1767, in get_optimizer
    optimizer = model.get_optimizer()
  File "/content/TTS/TTS/tts/models/vits.py", line 1643, in get_optimizer
    optimizer0 = get_optimizer(self.config.optimizer, self.config.optimizer_params, self.config.lr_disc, self.disc)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1269, in __getattr__
    raise AttributeError("'{}' object has no attribute '{}'".format(
AttributeError: 'Vits' object has no attribute 'disc'

**Reply by MaxIakovliev** [reply]
same issue here. Still persists. 
Any recommendations  on how to fix is much appreciated!

**Reply by rmcpantoja** [reply]
> same issue here. Still persists. Any recommendations on how to fix is much appreciated!

Hi @MaxIakovliev, 
I've fixed it by setting init_discriminator = True in the config.

**Reply by MaxIakovliev** [reply]
@rmcpantoja  - it works! Thanks

**Comment by francois-vz** [comment]
Could still be due to versioning.

**Reply by rmcpantoja** [reply]
> Could still be due to versioning.

Hi @Fransie-dev Yes, I have rolled back the commits to versions 0.8, 0.7.1 and 0.7 and the problem persists. I could change to a minor version, but the question is that it is not known with which commit that model was trained.
I could train a model from scratch with the current version of 🐸TTS, but I don't have a GPU, only the ones from the free colab where you also have a six hour limit.

**Reply by statscol** [reply]
the VitsArgs config in which this model was trained has some parameter names different to the current version of TTS.


tts_models/es/css10/vits default config:

`VitsArgs(num_chars=129, out_channels=513, spec_segment_size=32, hidden_channels=192, hidden_channels_ffn_text_encoder=768, num_heads_text_encoder=2, num_layers_text_encoder=6, kernel_size_text_encoder=3, dropout_p_text_encoder=0.1, dropout_p_duration_predictor=0.5, kernel_size_posterior_encoder=5, dilation_rate_posterior_encoder=1, num_layers_posterior_encoder=16, kernel_size_flow=5, dilation_rate_flow=1, num_layers_flow=4, resblock_type_decoder='2', resblock_kernel_sizes_decoder=[3, 5, 7], resblock_dilation_sizes_decoder=[[1, 2], [2, 6], [3, 12]], upsample_rates_decoder=[8, 8, 4], upsample_initial_channel_decoder=256, upsample_kernel_sizes_decoder=[16, 16, 8], periods_multi_period_discriminator=[2, 3, 5, 7, 11], use_sdp=True, noise_scale=1.0, inference_noise_scale=0.3, length_scale=1.0, noise_scale_dp=1.0, inference_noise_scale_dp=0.3, max_inference_len=None, init_discriminator=False, use_spectral_norm_disriminator=False, use_speaker_embedding=True, num_speakers=0, speakers_file='/root/.local/share/tts/tts_models--es--css10--vits/speaker_ids.json', d_vector_file=None, speaker_embedding_channels=256, use_d_vector_file=False, d_vector_dim=0, detach_dp_input=True, use_language_embedding=True, embedded_language_dim=4, num_languages=1, language_ids_file=None, use_speaker_encoder_as_loss=False, speaker_encoder_config_path='', speaker_encoder_model_path='', condition_dp_on_speaker=True, freeze_encoder=False, freeze_DP=False, freeze_PE=False, freeze_flow_decoder=False, freeze_waveform_decoder=False, encoder_sample_rate=None, interpolate_z=True, reinit_DP=False, reinit_text_encoder=False)`


using the same args with an updated version of TTS

`VitsArgs(num_chars=129, out_channels=513, spec_segment_size=32, hidden_channels=192, hidden_channels_ffn_text_encoder=768, num_heads_text_encoder=2, num_layers_text_encoder=6, kernel_size_text_encoder=3, dropout_p_text_encoder=0.1, dropout_p_duration_predictor=0.5, kernel_size_posterior_encoder=5, dilation_rate_posterior_encoder=1, num_layers_posterior_encoder=16, kernel_size_flow=5, dilation_rate_flow=1, num_layers_flow=4, resblock_type_decoder='2', resblock_kernel_sizes_decoder=[3, 5, 7], resblock_dilation_sizes_decoder=[[1, 2], [2, 6], [3, 12]], upsample_rates_decoder=[8, 8, 4], upsample_initial_channel_decoder=256, upsample_kernel_sizes_decoder=[16, 16, 8], periods_multi_period_discriminator=[2, 3, 5, 7, 11], use_sdp=True, noise_scale=1.0, inference_noise_scale=0.3, length_scale=1, noise_scale_dp=1.0, inference_noise_scale_dp=1.0, max_inference_len=None, init_discriminator=True, use_spectral_norm_disriminator=False, use_speaker_embedding=True, num_speakers=0, speakers_file='/root/.local/share/tts/tts_models--es--css10--vits/speaker_ids.json', d_vector_file=None, speaker_embedding_channels=256, use_d_vector_file=False, d_vector_dim=0, detach_dp_input=True, use_language_embedding=True, embedded_language_dim=4, num_languages=1, language_ids_file=None, use_speaker_encoder_as_loss=False, speaker_encoder_config_path='', speaker_encoder_model_path='', condition_dp_on_speaker=True, freeze_encoder=False, freeze_DP=False, freeze_PE=False, freeze_flow_decoder=False, freeze_waveform_decoder=False, encoder_sample_rate=None, interpolate_z=True, reinit_DP=False, reinit_text_encoder=False)`

This is how i managed to fine-tune the model.

**Comment by rmcpantoja** [comment]
Hi @statscol , so that was the problem? So, don't you necessarily need the discriminator of the pretrained model to fine tuning? Anyway, thanks a lot for the comparison. I'll try to train a Spanish ljspeech dataset in a few hours and i will post the results.

**Comment by nicemanis** [comment]
I found out that there are differences in layer names between versions:
![image](https://github.com/coqui-ai/TTS/assets/9463816/c642d640-430e-4121-99d3-1b90d3df3ae3)

Maybe fixing layer names resolves the issue

**Comment by youness-658** [comment]
i m trying to fine toning the coqui xtts v2 but i get this ereur : Traceback (most recent call last):
  File "C:\Users\zh\Desktop\project ai\TTS\TTS\bin\train_tts.py", line 71, in <module>
    main()
  File "C:\Users\zh\Desktop\project ai\TTS\TTS\bin\train_tts.py", line 58, in main
    trainer = Trainer(
  File "C:\Users\zh\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\trainer\trainer.py", line 506, in __init__
    self.criterion = self.get_criterion(self.model)
  File "C:\Users\zh\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\trainer\trainer.py", line 2098, in get_criterion
    criterion = model.get_criterion()
  File "C:\Users\zh\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\torch\nn\modules\module.py", line 1940, in __getattr__
    raise AttributeError(
AttributeError: 'Xtts' object has no attribute 'get_criterion'  

anyone can help please

**Reply by eginhard** [reply]
This should be fixed in [our fork](https://github.com/idiap/coqui-ai-TTS) (available via `pip install coqui-tts`). This repository is not maintained anymore.
